### DISCUSSION TITLE
Translating error messages with i18n

### DISCUSSION THREAD

**Discussion by iamacook** [react-hook-form/react-hook-form#2632] [Q&A]
I'm using i18next to translate our site and the user can switch between languages. However, RHF's error messages don't translate when toggling between languages.

I have tried calling `trigger()` and `control.reRender()` on language change but nothing helps, bar reloading the page. If you have any advice on how to force reload the validation schema, it would be greatly appreciated.

**Comment by bluebill1049** [comment]
provide a simple codesandbox please.

**Reply by iamacook** [reply]
[Here](https://codesandbox.io/s/react-hook-form-validationschema-v6-forked-9huqn?file=/src/index.js) is an example. Thank you in advance.

**Reply by bluebill1049** [reply]
i think you will need to build a custom resolver, and leverage context. we cache resolver inside hook form.

https://react-hook-form.com/api#useForm

<img width="983" alt="Screen Shot 2020-08-18 at 5 59 53 pm" src="https://user-images.githubusercontent.com/10513364/90486635-9fb73600-e17c-11ea-8e5f-26eadf523db3.png">

**Reply by iamacook** [reply]
Ok, thank you.

**Reply by sushmitg** [reply]
@iamacook Did you find the solution?

**Reply by iamacook** [reply]
@sushmitg Sadly not I'm afraid

**Reply by bluebill1049** [reply]
I am working on v7, what would be the idea API for that?

**Reply by iamacook** [reply]
Really appreciate you taking the time to consider this. As far as I can tell, there are only two viable solutions:

1. Using the `i18n.on('languageChanged', () => {})` event to reload the validation schema. However, RHF would need access to the i18n instance to listen for this.
2. Or as you previously suggested, build a custom resolver that uses the context.

The second is on my todo list, but it is currently not a pressing issue.

(I did come up with a horrible work around, however, in case anyone needs a temporary solution: save the form values on the `languageChanged` event, reload the page, import saved values as `defaultValues` and `trigger()`.)

**Reply by bluebill1049** [reply]
is any way we can do it within the `register` function?

**Reply by iamacook** [reply]
I mean, you could load the i18n instance within `register` but the language change affects the entire app and not individual fields. It would make more sense as a `useForm` attribute in my opinion.

**Reply by bluebill1049** [reply]
if all fields get re-register again, they r fine right? last option will be add more config to the `useForm`.

**Reply by iamacook** [reply]
I didn't think of that! In theory, yes. I can try to test it later and get back to you.

**Reply by iamacook** [reply]
I tried a few things out to no avail. [Here](https://codesandbox.io/s/react-hook-form-validationschema-v6-forked-jz70k?file=/src/index.js)'s an example where I un/re-register a field when the language changes, but the error message remains the same.

**Reply by bluebill1049** [reply]
i see... thanks for testing it out.

**Reply by ghost** [reply]
I found this solution: [here](https://codesandbox.io/s/react-hook-form-validationschema-v6-forked-jjhko?file=/src/index.js)
- error message just is the i18n code avoid i18n.t return a fixed message
- using i18n.t to translate inside the rendering body
- need a state/sth to trigger re-render

**Reply by FahadAly05** [reply]
@bluebill1049 did you found any solution?

**Reply by iamacook** [reply]
@FahadAly05 I got this working by using @st-quocpham's suggestion. Instead of returning the actual translation as the error, return the translation key instead:

```js
setError('fieldName', {
  type: 'manual',
  message: 'error.message.key'
});
```

**Reply by FahadAly05** [reply]
@iamacook  Thanks for the solution. But this method always called on language change even without submitting the form, and produce errors. Can you provide codesandbox link.

**Reply by iamacook** [reply]
[Here](https://codesandbox.io/s/react-hook-form-validationschema-v6-forked-pwp2m?file=/src/index.js) is an example. You need add the `i18n.t` function arround the key. In our case we created our own custom component that takes the input `name` via props to access its specific error message: `get(errors, props.name)`. You can do something like this:

```ts
const getErrorMessage = (name: string, errors: DeepMap<Record<string, any>, FieldError>): undefined | string => {
  const message = get(errors, name)?.message
  if (!message) return // No error

  // Check if dictionary entry exists, else return 'key'
  return i18n.exists(message) ? i18n.t(message) : message
}
```

**Reply by jukkahuuskonen** [reply]
I'm a little late on this matter, but here is my solution.

**ErrorMessage**-component
```
import styled from '@emotion/styled';
import useFormattedMessage from 'hooks/useFormatMessage';

const ErrorDiv = styled('div', {
  shouldForwardProp: prop => !['width', 'maxWidth'].includes(prop),
})`
  color: ${({ theme }) => theme.palette.error.main};
  padding: 1px 10px;
  max-width: ${({ width = 'auto' }) => width};
  max-width: ${({ maxWidth = 'none' }) => maxWidth};
  overflow: hidden;
`;

const ErrorMessage = ({ message, translate = true }) => {
  const { ucFirst } = useFormattedMessage();
  if (!message) {
    return null;
  }
  const finalMessage = translate ? ucFirst({ id: message }) : message;
  return <ErrorDiv>{finalMessage}</ErrorDiv>;
};

export default ErrorMessage;

```

**useFormattedMessage ** is my custom hook that gives several INTL-functions that change when locale changes.
**ucFirst** is same as **formatMessage** + uppercase first letter.
```
import { useCallback } from 'react';
import { useIntl } from 'react-intl';
import { ucFirst, titleCase } from 'change-case';

const useFormatMessage = () => {
  const { formatMessage, locale } = useIntl();
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const uppercaseFirst = useCallback(fmObject => ucFirst(formatMessage(fmObject)), [
    formatMessage,
    locale,
  ]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const capitalizeWords = useCallback(fmObject => titleCase(formatMessage(fmObject)), [
    formatMessage,
    locale,
  ]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const formatMessageWithLocale = useCallback(fmObject => formatMessage(fmObject), [
    formatMessage,
    locale,
  ]);
  return {
    formatMessage: formatMessageWithLocale,
    ucFirst: uppercaseFirst,
    capitalizeWords,
    locale,
  };
};

export default useFormatMessage;
```

Then you just pass the correct message to the **ErrorMessage**

```
<ErrorMessage message={<YOUR ERROR INTL ERRORKEY>} />
```

And you can use the same **ErrorMessage** with prop **translation={false}** if you don't need translation.

**Reply by bluebill1049** [reply]
thank you for sharing solution.

**Reply by mleimer** [reply]
The suggested workaround isn't a viable option in my opinion. It solves the problem of showing the error message in a different language upon a language change, but only as long as no additional arguments need to be passed into the translation-function later on. If additional parameters need to be passed, for example to describe a dynamic length limit, then the only option I can see so far is to stringify the object to comply with `Message` being of type `string`. Having said that, there are also no sub-classing possibilities. Moreover, having dynamic keys prevents the static discovery of translation keys for an export into a TMS or alike.

**Reply by shaungrady** [reply]
I've [opened a new issue](https://github.com/react-hook-form/react-hook-form/issues/10482) proposing a flexible solution that has worked for my team.

**Reply by npearson72** [reply]
@shaungrady and @bluebill1049 what ever happened to @shaungrady's PR? Is there another solution available?

**Reply by shaungrady** [reply]
Sorry, I haven't had time to devote to this.

**Comment by terryatgithub** [comment]
Is there some official solution or progress about this issue? we also met this problem. 
any feedback would be appreciated.

**Comment by debel27** [comment]
Here's my proposal to solve the issue. 

@bluebill1049 could you tell me what you think? If you see no red flag, I can attempt a PR.

### Main idea

This proposal aims at giving developers more flexibility when defining errors objects in react-hook-form.

The main idea is to add a `data` property to `FieldError`, `ErrorOption` and `GlobalError`. Example:

```typescript
export type FieldError = {
  // ... existing properties
  message?: Message;
  // -- new property --
  data?: unknown;
};
```

RHF will treat the new property as opaque and just forward it along.

In application space, developers can pick whichever information they need from `FieldError` to display the error properly. They can keep using `message` as usual, or they can use `data` for more advanced use cases.

### Producing `data`

Like `message`, the `data` property will be filled in by validators.

For declarative validators, we can add a `data` property to `ValidationValueMessage`:

```typescript
export type ValidationValueMessage<TValidationValue extends ValidationValue = ValidationValue,> = {
  value: TValidationValue;
  message: Message;
  // -- new property --
  data?: unknown;
};
```

For function validators, we will extend `ValidateResult` to support arbitrary data:

```typescript
export type ValidateResult = 
  | Message
  | Message[]
  | boolean
  | { data: unknown } // <- new union member
  | undefined;
```

In a nutshell, the goal is to provide first-class support for `data` in every place `Message` is used.

In resolver functions, the problem of providing `data` will be solved in userland.

### Caveat: HTML validation

The proposal works fine when error objects are used as **reactive data** ([as defined by React](https://react.dev/learn/separating-events-from-effects#reactive-values-and-reactive-logic)), because RHF can delegate error formatting to application space.

However, if RHF is configured to use native HTML validation, errors will be used as **event data** instead. RHF will imperatively call `HTMLInputElement.setCustomValidity` whenever an error occurs, and must therefore produce a localized message on-demand.

This raises the following questions:

  1. given a `FieldError` message, how can the library decide whether to use `message` or `data`?
  2. if `data` is picked, how to transform it into a `string` for `setCustomValidity`?

`1.` can be solved by picking `message` first (as in `message ?? data`).

`2.` requires RHF to operate in "inversion of control" mode. Developers can provide a `formatErrorData` function to the library (`(key: string, data: unknown) => string`), which will be called to derive `message` from `data` in case the former is absent. If the function is not provided, `data` will be ignored in HTML validation.
